index: remove unnecessary locator cleaning in BaseIndex::Init()

HowHsu opened the pull request:
Leverage `locator.IsNull()` to simplify `ReadBestBlock()` and remove
unnecessary `SetNull()`.

l0rinc reviewed (approved):
ACK 1f678d832b1196ca7d7c2ff7e70d2172ae8c90b6

Recreated and retested it locally. It's a similar change to https://github.com/bitcoin/bitcoin/pull/33042, we have a few more places which should be cleanup up like this.

I think this already makes the code slightly simpler, but alternatively, you could return an optional and not rely output params and nullness

<details>
<summary>Details</summary>

```patch
diff --git a/src/index/base.cpp b/src/index/base.cpp
index fdd0e0d8af..0408021fd5 100644
--- a/src/index/base.cpp
+++ b/src/index/base.cpp
@@ -59,13 +59,10 @@ BaseIndex::DB::DB(const fs::path& path, size_t n_cache_size, bool f_memory, bool
         .options = [] { DBOptions options; node::ReadDatabaseArgs(gArgs, options); return options; }()}}
 {}
 
-bool BaseIndex::DB::ReadBestBlock(CBlockLocator& locator) const
+std::optional<CBlockLocator> BaseIndex::DB::ReadBestBlock() const
 {
-    bool success = Read(DB_BEST_BLOCK, locator);
-    if (!success) {
-        locator.SetNull();
-    }
-    return success;
+    if (CBlockLocator locator; Read(DB_BEST_BLOCK, locator)) return locator;
+    return {};
 }
 
 void BaseIndex::DB::WriteBestBlock(CDBBatch& batch, const CBlockLocator& locator)
@@ -97,20 +94,17 @@ bool BaseIndex::Init()
     // callbacks are not missed once m_synced is true.
     m_chain->context()->validation_signals->RegisterValidationInterface(this);
 
-    CBlockLocator locator;
-    if (!GetDB().ReadBestBlock(locator)) {
-        locator.SetNull();
-    }
+    const auto locator{GetDB().ReadBestBlock()};
 
     LOCK(cs_main);
     CChain& index_chain = m_chainstate->m_chain;
 
-    if (locator.IsNull()) {
+    if (!locator) {
         SetBestBlockIndex(nullptr);
     } else {
         // Setting the best block to the locator's top block. If it is not part of the
         // best chain, we will rewind to the fork point during index sync
-        const CBlockIndex* locator_index{m_chainstate->m_blockman.LookupBlockIndex(locator.vHave.at(0))};
+        const CBlockIndex* locator_index{m_chainstate->m_blockman.LookupBlockIndex(locator->vHave.at(0))};
         if (!locator_index) {
             return InitError(Untranslated(strprintf("best block of %s not found. Please rebuild the index.", GetName())));
         }
diff --git a/src/index/base.h b/src/index/base.h
index 4131b06cad..c1971cae3f 100644
--- a/src/index/base.h
+++ b/src/index/base.h
@@ -56,7 +56,7 @@ protected:
            bool f_memory = false, bool f_wipe = false, bool f_obfuscate = false);
 
         /// Read block locator of the chain that the index is in sync with.
-        bool ReadBestBlock(CBlockLocator& locator) const;
+        std::optional<CBlockLocator> ReadBestBlock() const;
 
         /// Write block locator of the chain that the index is in sync with.
         void WriteBestBlock(CDBBatch& batch, const CBlockLocator& locator);
```

</details>

HowHsu committed:
refactor: remove redundant locator cleanup in BaseIndex::Init()

Leverage locator.IsNull() to simplify ReadBestBlock() and remove
unnecessary SetNull().

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>
Co-authored-by: l0rinc <pap.lorinc@gmail.com>
Signed-off-by: Hao Xu <hao.xu@linux.dev>

l0rinc commented:
There's still a typo in the title

l0rinc reviewed (approved):
ACK ce7323ec635c389732e142c043f83868ddcefdb3

The commit message still sounds a bit unnatural and has a few typos, consider something like:
> index: remove unnecessary locator cleaning in BaseIndex::Init()
>
> Refactor BaseIndex::DB::ReadBestBlock() to return an optional value so that
callers don't need to rely on the locator itself to check if the call is
successful or not. This simplifies the code.

or just simply:
> refactor: remove redundant locator cleanup in BaseIndex::Init()
>
> Return optional value in `ReadBestBlock()` to eliminate manual `SetNull()` calls.

l0rinc commented:
ACK 04044e554e6982977f9d82f9bb60659315a45f92

PR title still has the typo

furszy commented on src/index/base.cpp:
```diff
@@ -59,13 +59,10 @@ BaseIndex::DB::DB(const fs::path& path, size_t n_cache_size, bool f_memory, bool
-    }
-    return success;
+    if (CBlockLocator locator; Read(DB_BEST_BLOCK, locator)) return locator;
+    return {};
```
Better to `return std::nullopt`.

l0rinc commented:
ACK c83741e4033824b4c798fc3e9cdc33cb30bbaf32

please fix the typo in the PR subject: locater -> locator

l0rinc commented on src/index/base.cpp:
```diff
@@ -97,20 +94,17 @@ bool BaseIndex::Init()
     CChain& index_chain = m_chainstate->m_chain;
 
-    if (locator.IsNull()) {
+    if (!locator) {
```
Since there's already a null locator, you're right, we don't need an optional

l0rinc commented:
ACK facd01e6ffbbd019312f370a3807de0b95bbd745

Note that we don't usually add the `Signed off by` to the commit message, it just seems like noise to me. You can still sign your commits if you want, but I'm no sure what this line adds that isn't already obvious.

Also, not sure why `refactor: remove redundant locator cleanup in BaseIndex::Init()` is repeated (slightly differently) in the PR description.

furszy reviewed (commented):
utACK facd01e6ffbbd019312f370a3807de0b95bbd745

At least for me, there’s no need to add me as a co-author when the code or suggestion is simple. Generally, only add people when they contribute a substantial change.
Just a small thought.

sedited reviewed (approved):
ACK facd01e6ffbbd019312f370a3807de0b95bbd745
